packages/back-end/src/services/features.ts

```diff
@@ -590,3 +649,139 @@ export function getNextScheduledUpdate(
+          return rule;
+        });
+      }
+      acc[key] = feature;
+      return acc;
+    },
+    {}
+  );
+}
+
+export function applyExperimentHashing(
+  experiments: SDKExperiment[],
+  attributes: SDKAttributeSchema,
+  salt: string
+): SDKExperiment[] {
+  return experiments.map((experiment) => {
+    if (experiment?.condition) {
+      experiment.condition = applyConditionHashing(
+        experiment.condition,
+        attributes,
+        salt
+      );
+    }
+    return experiment;
+  });
+}
+
+export function applyConditionHashing(
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  condition: any,
+  attributes: SDKAttributeSchema,
+  salt: string
+) {
+  return hashStrings({ obj: condition, salt, attributes });
+}
+
+interface hashStringsArgs {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  obj: any;
+  attribute?: SDKAttribute;
+  doHash?: boolean;
+}
+function hashStrings({
+  obj,
+  salt,
+  attributes,
+  attribute,
+  doHash = false,
+}: hashStringsArgs & {
+  salt: string;
+  attributes: SDKAttributeSchema;
+}): // eslint-disable-next-line @typescript-eslint/no-explicit-any
+any {
+  if (Array.isArray(obj)) {
+    // parse array
+    const newObj = [];
+    for (let i = 0; i < obj.length; i++) {
+      newObj[i] = processVal({
+        obj: obj[i],
+        attribute,
+        doHash,
+      });
+    }
+    return newObj;
+  } else if (typeof obj === "object" && obj !== null) {
+    // parse object
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const newObj: any = {};
+    for (const key in obj) {
+      // check if a new attribute is referenced, and whether we need to hash it
+      // otherwise, inherit the previous attribute and hashing status
+      attribute = attributes.find((a) => a.property === key) ?? attribute;
+      doHash = attribute
+        ? !!(
+            attribute?.datatype &&
+            ["secureString", "secureString[]"].includes(
+              attribute?.datatype ?? ""
+            )
+          )
+        : doHash;
+
+      newObj[key] = processVal({
+        obj: obj[key],
+        attribute,
+        doHash,
+      });
+    }
+    return newObj;
+  } else {
+    return obj;
+  }
+
+  function processVal({
+    obj,
+    attribute,
+    doHash = false,
+  }: hashStringsArgs): // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  any {
+    if (Array.isArray(obj)) {
+      // recurse array
+      return hashStrings({ obj, salt, attributes, attribute, doHash });
+    } else if (typeof obj === "object" && obj !== null) {
+      // recurse object
+      return hashStrings({ obj, salt, attributes, attribute, doHash });
+    } else if (typeof obj === "string") {
+      // hash string value
+      return doHash ? sha256(obj, salt) : obj;
+    } else {
+      return obj;
+    }
+  }
+}
+
+function sha256(str: string, salt: string): string {
+  return createHash("sha256")
+    .update(salt + str)
+    .digest("hex");
+}
```

Comment:
Use HMAC with SHA-256 instead of trying to prepend a salt: https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/sign#hmac_2

Salts are by definition random and not secret, used to prevent attacks based on the pre-computation of values. What you're using is a consistent value that is a secret.

If I read the context of this PR correctly, what you're trying to implement is a blind index.

Reply:
Also be aware that attributes with binary or otherwise low cardinality values can still leak information as their outputs will all be the same.

Reply:
Because this is client-executed hashing with a public salt, the salt really only serves the purpose of obfuscation and does not provide any significant protection from a determined hacker building rainbow tables, etc. As such, I think that sha256(input + publicSalt) is sufficient and easier to implement across different languages (this operation needs to be performed across all sdk environments, not just JS).

We are also planning on building out a server-side feature evaluation mechanism that provides actual privacy. We'd strongly recommend exclusively using server side evaluation (or just server side SDKs) where security/privacy of payload matters.

Reply:
I thought that being in the "back-end" folder meant that this was server-side code and therefore able to afford a level of privacy. Am I reading that wrong?

Reply:
This is the GB app's backend code that prepares a hashed payload which the frontend will later use for comparison. So while the initial hashing is backend, the comparison happens in an insecure environment.

Reply:
I added some additional information about hashing and encryption, and warned about security shortcomings, within the GB UI

Reply:
I'm still uneasy with this, but at least it's documented. Thank you.